### PR TITLE
Front-end support for arrays of unboxed products

### PR DESCRIPTION
**Pull request by ccasin**
This adds the front-end bits of support for arrays of unboxed products. In particular:
- It adds two new array kinds - one for arrays of products where everything is GC scannable, and another for arrays of products where everything is GC ignorable.
- It adds a new primitive (which has no native code implementation yet) for creating arrays, `"%makearray_dynamic"`, so named because it is like `makearray` but the size is not statically known.

This is all limited to alpha because there is no middle- or back-end support for these things yet. There are typing tests showing that all the array primitives work on product arrays but reject products that aren't all GC scannable or all GC ignorable.  It doesn't yet add the planned "re-interpret" primitives or "size of". Those will come soon, but there is still discussion about the exact design.

Note that the new array creation primitive isn't specific to products - it can be used to create arrays containing data of any representable layout. Our plan is to eventually replace the per-layout C functions with this primitive.

There are no dynamic tests here. Tests will come in the follow-up PR that adds middle- and back-end support, which will also move these to beta. I actually think the bytecode implementation here works, and have started testing it over in [this branch](https://github.com/ccasin/flambda-backend/tree/tuple-arrays-frontend-test). But I can't easily include those tests here because the fact that "any" is interpreted differently in alpha makes it impossible to use our generic unboxed array testing framework while these arrays are in alpha.

**Review:** @goldfirere.
- Note there are three commits.  The first two are just #3138, so only the last needs review here.
- The odd looking diff in `typing/primitive.ml` is a result of me accidentally including the typing rule for the new primitive prematurely in #3092, but with the name we planned when we thought it would only work with product arrays.

### CODE REVIEW

**Comment by ccasin** [comment]
Note this now also adds a new `"%arrayblit"` primitive, gated to alpha, that will be implemented by a future backend PR

**Comment by ccasin** [comment]
[This commit](https://github.com/ccasin/flambda-backend/commit/a55f456a2d1688848f9565621e35dca696ceeebb) adds dynamic tests for everything added in this PR.  It passes in bytecode.  I can't include it in this PR because the fact that `any` means something different in alpha makes it hard to used the arrays-of-unboxed-things testing infrastructure.  It will be included with the PR that adds native code implementations and moves this all to beta.

**Review by goldfirere — COMMENTED** [review]
Reviewed tests (only). More in due course.

File: ocaml/testsuite/tests/typing-layouts-products/basics.ml
Comment (ccasin) [reply ocaml/testsuite/tests/typing-layouts-products/basics.ml]: Revised, thanks.

File: ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml
Comment (ccasin) [reply ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml]: Well spotted!  Fixed.

File: ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml
Comment (ccasin) [reply ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml:128]: We resolved this in person: In principle we could allow such a thing, but the implementation of `layout_poly` externals makes it tricky, and it doesn't seem worth complicating things to suppport since you can't make anything of these types anyway.

File: ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml
Comment (ccasin) [reply ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml:1]: I think this is not worth testing.  There is no checking that externals are given sensible types - you can freely bind them at nonsense types and get unspecified behavior.  On the other hand, there is checking that externals are given types with sensible layouts, and that is exercised here.

Comment (ccasin) [reply ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml:1]: I added a new test for this at the end, rather than integrating it with existing ones.  The existing tests demonstrate that typing is the same for all the relevant primitives sufficiently that I'm content to just test this on one primitive, but let me know if you disagree.

(If these tests were going to stick around forever, I would have spent the time to make them generated in such a way that they are easy to change uniformly.  But in fact they will go away when we support arrays of arbitrary unboxed products, which is planned for "soon".)

Comment (ccasin) [reply ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml:1]: Added a test for this.

Comment (ccasin) [reply ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml:1]: Agreed.  I'll just point out that the fact that the layout checking sometimes happens later than we'd like isn't new to this PR: you can already observe behavior like this due to the nature of the `[@layout_poly]` implementation.

File: ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml
Comment (ccasin) [reply ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml:1824]: I think this falls into the category of giving a primitive an incorrect type, and is thus just unsupported.

File: ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml
Comment (ccasin) [reply ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml:1825]: I agree they are going to want such things, but think we should leave it for a future PR as it will involve either substantially generalizing or duplicating the two new magic externals (`%makearray_dynamic` and `%arrayblit`) and their implementations in the middle and backend are going to be complicated enough already.

(Also I think 'returned to the manufacturer' probably overstates it: they will be annoyed at having to add some extra `Int64_u.to_int`s, but it's not a fundamental obstacle)

File: ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml
Comment (ccasin) [reply ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml:1]: Good point.  I added the first four here.

In testing the last one, I found we have no existing tests for unboxed types in comprehensions and in fact encountered a pre-existing bug. This is now fixed in #3190, which I think adequately tests the interaction of comprehensions and non-value layouts.

File: ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml
Comment (ccasin) [reply ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml:1825]: (I've made a ticket for this)

File: ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml
Comment (ccasin) [reply ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml:1]: Good question. I took a quick look around the testsuite and as far as I can tell there are no tests at all for the interaction of immutable arrays and unboxed types. Which seems bad!  But fixing that is probably beyond the scope of this PR.

For now, I've pushed a commit that bans use of unboxed products in immutable arrays (and a corresponding test).  I expect this will not be tricky to add, but considering the number of moving pieces here I agree it makes sense to do in a future PR.

**Comment by ccasin** [comment]
@goldfirere back at you

File: ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml
Comment (goldfirere) [reply ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml:1825]: I guess maybe I'm wondering if we need the tagged primitives at all. Maybe we just offer the untagged ones, and implement any desired tagged version in terms of an untagged primitive.

File: ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml
Comment (goldfirere) [reply ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml:1]: Strong agree on "separate PR". Thanks.

File: ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml
Comment (goldfirere) [inline ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml:2028]: Good tests for literals here. I think we thought it might be helpful to leave a CR to make sure these are also tested in the middle/back end when that becomes possible.

**Review by goldfirere — COMMENTED** [review]
Timing out for now. More soon.

File: ocaml/typing/typeopt.ml
Comment (ccasin) [reply ocaml/typing/typeopt.ml:194]: > Indeed. This seems kinda bad.

I think there is an interesting discussion to be had here, but it's probably worth saying explicitly that that discussion shouldn't hold up this PR. The thing implemented here is fine for a first version.

> We could, of course, introduce product kinds. But I continue to think this is the wrong idea -- and it will only get worse, once kinds have `with` things in them.
> 
> What I want is a way to store both a sort and a gc-ignore-flag (gif) for each component of a product. But if we have a subsumption relationship between can-ignore and cannot-ignore, then we lose our nice property of having a discrete set of incomparable sorts. After thinking about this a little while, I'm stumped. Let's talk about this.

I remain confused why we wouldn't want product kinds. The more time we spend working around their absence, the more I think we should at least give them a try. It's definitely an interesting observation that we can work around not having them, but I don't think they complicate inference, and in fact I think they will just make the whole system simpler.

I'm certainly not proposing to introduce them right now, though - we should get this and a first version of unboxed records and interior pointers out with the system we have.

(Speaking of making the whole system simpler, I think can even imagine a world where product kinds allow us to eliminate product both product layouts and product sorts. Though I think they'd simplify the system even if not.)

> (To be clear, even if we had product kinds, I think the inference problem lurks. So it's just not as simple as "add product kinds".)

This is not clear to me. I think product kinds would lead to a straightforward implementation of the product cases of `array_kind_of_elt`, working only on the kind.  We would (1) classify each element of the product by its layout and ignorability, (2) check we don't have a mixed block, and (3) return the obvious array kind.

> Actually, I've thought about this more (longer than I should admit, but in the evening at least).
> 
> I think trying to infer stuff in this space is the wrong thing to be trying for. We don't, for example, try to eagerly figure out where we can specialize types to `int` so that typeopt can do its thing. Instead, here in typeopt, we look deeply through types to figure out all we can.

I agree that typeopt looks deeply through types.

I am not convinced this is something to emulate, though. `value_kind` is a significant source of both complexity and performance problems, exactly because of this. I continue to think there should be one function that answers the question "what is the kind of this type", and that function is `estimate_type_jkind`.

I can actually imagine wanting to go further the other way - fold more of what `value_kind` is looking for into the kind system, until its implementation too can just look at the kind.

> (EDIT: The next bit of text probably is more about @rtjoa's unboxed records than this current patch, but it all comes back together in the end, and I know you @ccasin care deeply about both patches.)
> 
> And, thinking about "look deeply through types" made me realize what the specification of `scrape_ty` is: `scrape_ty env ty` returns a `ty'` whose runtime representation is the same as `ty` (but is somehow more primitive / easier to understand).

I agree that's a reasonable specification of `scrape_ty`. `scrape_ty` is just calls to `expand_head_opt` and `get_unboxed_type_approximation`, and this is the specification of those functions (or, at least, that has been my interpretation of the somewhat cryptic comment about "type-based optimizations" on `expand_head_opt`).

> This suggests that we should add a new ability to `scrape_ty`: "look through" an unboxed record to get an unboxed tuple. This meets the specification of `scrape_ty` beautifully: an unboxed tuple shares a runtime representation with an unboxed record, and the tuple is easier to process. (Yes, I know I just introduced that specification for `scrape_ty` and then used it -- but I really do think this is the useful specification of that function.)

You may recall from an earlier conversation that @rtjoa's first prototype for unboxed records took this approach (having `get_unboxed_type_representation` return an unboxed tuple when it sees an unboxed record).  I think this is a fun idea, but doesn't work here because of mutability. Note that `type_opt` cares about mutability of record fields in `value_kind`, and I expect this approach would make it hard to implement that logic when we allow for nested unboxed records in other structures.

(One could imagine enhancing tuple types to add a notion of mutable field just for this purpose, but for the reasons above I'm not particularly inclined to go this route.)

> We can go further. I conjecture that, today, `scrape_ty` and `get_unboxed_type_approximation` always do the same thing (though the latter does it with a little less work). (Modulo correcting levels, of course. That step, happily, does not need to be done once we merge with 5.3.)

It's a nice observation that `scrape_ty` and `get_unboxed_type_approximation` have pretty similar implementations and could potentially be just one function.  I think that might be right.

Comment (goldfirere) [reply ocaml/typing/typeopt.ml:194]: > I think there is an interesting discussion to be had here, but it's probably worth saying explicitly that that discussion shouldn't hold up this PR. The thing implemented here is fine for a first version.

Strong agree.

> You may recall from an earlier conversation that @rtjoa's first prototype for unboxed records took this approach (having get_unboxed_type_representation return an unboxed tuple when it sees an unboxed record).

Yes, I do! And I should have noted this in my post; sorry for the oversight.

As for the rest: this will be much better discussed live than on GitHub. It would be reasonable to delay that discussion until after this all lands, but I think it's all fresh now and worth at least agreeing on next steps here. I'll suggest a time, but I also understand if we want to delay.

**Review by goldfirere — APPROVED** [review]

File: ocaml/typing/primitive.ml
Comment (mshinwell) [reply ocaml/typing/primitive.ml:651]: The thing that will be called `%makearray_dynamic`.

Comment (ccasin) [reply ocaml/typing/primitive.ml:651]: Indeed, to quote an (easily missed) note in the PR description:

> The odd looking diff in typing/primitive.ml is a result of me accidentally including the typing rule for the new primitive prematurely in https://github.com/ocaml-flambda/flambda-backend/pull/3092, but with the name we planned when we thought it would only work with product arrays.

File: ocaml/lambda/translprim.ml
Comment (ccasin) [reply ocaml/lambda/translprim.ml:1040]: Yes, this is the pre-existing cheat meantioned in the comment.  `Translprim.lookup_primitive` doesn't have access to type information and just always uses the `Pgenarray` versions of things, the these `glb` functions fix it up.

This is gross! I'd like to do a big rework of how primitives are handled in lambda and of the layout poly primitives in particular, but it's outside the scope of this PR.

File: ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml
Comment (ccasin) [reply ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml]: I think the primitives for tagged ints are essential.  `int` remains a pervasive and convenient to use type, and it's totally reasonable to use it with these arrays - saving allocations and indirections is more important than saving a bitshift in an indexing operation.  It's clear we're going to need all the others too, and will add them in due course.

File: middle_end/flambda2/from_lambda/lambda_to_flambda.ml
Comment (ccasin) [reply middle_end/flambda2/from_lambda/lambda_to_flambda.ml]: Good catch, I think this was just wrong and have moved it.  Will confirm with Mark.

File: ocaml/bytecomp/bytegen.ml
Comment (ccasin) [reply ocaml/bytecomp/bytegen.ml:1]: Yes, but not in this PR - these changes should be tested using our framework for testing arrays of unboxed types, and as noted elsewhere that's imcompatible with "alpha" because of the difference in meaning of "any".  The commit mentioned above with dynamic tests includes bytecode tests, and will be part of the PR that moves this all out of beta.

File: ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml
Comment (goldfirere) [reply ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml]: I'm fine with leaving this Resolved, but it seems simpler to me to implement a primitive over untagged things and then just have a little wrapper around it for tagged ones. That is, I would want the tagged version available to users, as you suggest, but I don't see why that tagged version needs to be a primitive. Maybe it would be more efficient? But I would hope not, with our optimizer.

Comment (ccasin) [reply ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml]: Yes, sorry, I misread your comment.  I agree that potentially makes sense, and we could consider switching to it in a future PR that adds the untagged versions.

**Pull request merged**
